src/FreeAge/common/type_stats_data.cpp

```diff
@@ -0,0 +1,330 @@
+  s.size = QSize(0, 0);
+  s.occupancy = QRect(0, 0, 0, 0);
+}
+
+inline void SetBuildingBasic(BuildingTypeStats& s, int maxHp, int meleeArmor, int pierceArmor, int size, float lineOfSight) {
+  s.maxHp = maxHp;
+  s.armor.SetValue(DamageType::Melee, meleeArmor);
+  s.armor.SetValue(DamageType::Pierce, pierceArmor);
+  s.armor.SetValue(DamageType::Building, 0);
+  s.armor.SetValue(DamageType::StandardBuilding, 0);
+  s.size = QSize(size, size);
+  s.occupancy = QRect(0, 0, size, size);
+  s.lineOfSight = lineOfSight;
+}
+
+inline void SetBuildingGarrison(BuildingTypeStats& s, GarrisonType garrisonType, int garrisonCapacity) {
+  s.garrisonType = garrisonType;
+  s.garrisonCapacity = garrisonCapacity;
+}
+
+inline void SetBuildingAttack(BuildingTypeStats& s, AttackType attackType, float fireRate, int pierceDamage, float accuracy, float maxRange, float projectileSpeed, float attackDelay = 0) {
+  s.attackType = attackType;
+  s.projectileSpeed = projectileSpeed;
+  s.fireRate = fireRate;
+  s.damage.SetValue(DamageType::Pierce, pierceDamage);
+  s.accuracy = accuracy;
+  s.maxRange = maxRange;
+  s.projectileSpeed = projectileSpeed;
+  s.attackDelay = attackDelay;
+}
+
+inline void SetBuildingBonusDamage(BuildingTypeStats& s, DamageType damageType, int value) {
+  s.damage.SetValue(damageType, value);
+}
+
+inline void SetBuildingCost(BuildingTypeStats& s, float creationTime, u32 wood, u32 food, u32 gold, u32 stone) {
+  s.creationTime = creationTime;
+  s.cost = ResourceAmount(wood, food, gold, stone);
+}
+
+// Vector helpers
+
+inline UnitTypeStats& InitUnit(std::vector<UnitTypeStats>& unitTypeStats, UnitType type) {
+  UnitTypeStats& s = unitTypeStats[static_cast<int>(type)];
+  SetUnitDefaults(s);
+  return s;
+}
+inline UnitTypeStats& CopyUnit(std::vector<UnitTypeStats>& unitTypeStats, UnitType type, UnitType source) {
+  UnitTypeStats& s = unitTypeStats[static_cast<int>(type)];
+  s = unitTypeStats[static_cast<int>(source)];
+  return s;
+}
+
+inline BuildingTypeStats& InitBuilding(std::vector<BuildingTypeStats>& buildingTypeStats, BuildingType type) {
+  BuildingTypeStats& s = buildingTypeStats[static_cast<int>(type)];
+  SetBuildingDefaults(s);
+  return s;
+}
+inline BuildingTypeStats& CopyBuilding(std::vector<BuildingTypeStats>& buildingTypeStats, BuildingType type, BuildingType source) {
+  BuildingTypeStats& s = buildingTypeStats[static_cast<int>(type)];
+  s = buildingTypeStats[static_cast<int>(source)];
+  return s;
+}
+
+// Data
+
+// NOTE: The following functions should be able to inline all the helper function calls and elimenate
+//       any redundant value sets.
+
+void LoadUnitTypeStats(std::vector<UnitTypeStats>& unitTypeStats) {
+  // pre allocate all the UnitTypeStats and then fill the data
+  unitTypeStats.resize(static_cast<int>(UnitType::NumUnits));
+
+  { // Villager
+    UnitTypeStats& s = InitUnit(unitTypeStats, UnitType::FemaleVillager);
+
+    SetUnitBasic(s,
+      /* hp     */ 25,
```

Comment:
Personally, I don't like that the argument is only defined by the order. Looks like a great source of copy&paste errors when completing the unit stats. I have literally no experience with C++ so I don't know what options there are -- no named arguments as far as I can tell. 

Would a chained, object-oriented approach work? Sth like: 

```c++
s = InitUnit(unitTypeStats, UnitType::FemaleVillager)
          .setHp(25)
          .setArmor(0,0)
          .setRadius(1)
```

Reply:
This would work, but in contrast to a function like `SetUnitBasic()`, it would not directly ensure that all of the relevant options are set together (so it would be possible to forget setting a new option for some unit, which may then remain uninitialized). It would be possible to add a mechanism that checks for a complete initialization, but that would also add some complexity.

Reply:
You example could be implemented, but the reason of the functions is to ensure that some info is given grouped for example:
```c++
void SetUnitMeleeAttack(UnitTypeStats& s, 
  float fireRate, int meleeDamage, float attackDelay)
```
sets everything that is needed for a melee attack, and sets more things than the 3 given params. It also enusres eg. that if you have a melee atack you have explicity set the fireRate and not left the default value.

I understand that there is a risk that you reorder things believing its not a functions, but I think is risk worth taking.

Reply:
( I didn't want to repeat @puzzlepaint, his answer had not loaded when I wrote mine )

Reply:
Thinking about this, I came up with an idea for "simulating" named parameters in C++:

```c++
template <typename T>
class NamedParameterBase {
 public:
  inline NamedParameterBase(T value)
      : value(value) {}
  
  inline operator T() { return value; }

 private:
  T value;
};

#define MAKE_NAMED_PARAMETER(ParamName, ParamType) \
    class ParamName : public NamedParameterBase<ParamType> { using NamedParameterBase<ParamType>::NamedParameterBase; }

MAKE_NAMED_PARAMETER(FireRate, float);
MAKE_NAMED_PARAMETER(MeleeDamage, int);
MAKE_NAMED_PARAMETER(AttackDelay, float);

#undef MAKE_NAMED_PARAMETER

// Then one can define:
void SetUnitMeleeAttack(UnitTypeStats& s, 
  FireRate fireRate, MeleeDamage meleeDamage, AttackDelay attackDelay);

// ... which is later used like this:
SetUnitMeleeAttack(
    FireRate(0.4f),
    MeleeDamage(2),
    AttackDelay(0.3));
```

This does add some (but not much) code overhead. It might be an idea for formalizing the inline comments and prevents mixing up the order as suggested by @grst.

Reply:
You still cannot reorder the params but it will give an error, right? How you then get the value out, with `fireRate()` ?

Could this use [User-defined literals](https://en.cppreference.com/w/cpp/language/user_literal) ?

Reply:
Yes, this is just for having (and thus verifying) the parameter names in the function calls. it does not make reordering possible.

In the example code above, getting the value out would work with the conversion operator that is defined in NamedParameterBase. So, for a float parameter for example, in a context where the parameter is directly converted to a float, it could just be used like one. For example, this works: `float test = attackDelay;` This will not work though if the conversion is not clear, then something like `static_cast<float>(attackDelay)` would need to be used, I think. An alternative would be to make a simple `get()` function.

I haven't used user-defined literals before, but I think this should work with them as well, yes. It's probably just a question of which syntax seems preferable. Either:
```c++
SetUnitMeleeAttack(
    FireRate(0.4f),
    MeleeDamage(2),
    AttackDelay(0.3));
```
or:
```c++
SetUnitMeleeAttack(
    0.4_fireRate,
    2_meleeDamage,
    0.3_attackDelay);
```

Reply:
This is very clever, but I dont know... :)

```c++
    SetUnitBasic(s,
      /* max hp */ 40,
      /* armor  */ 0,  1,
      /* radius */ 1,
      /* speed  */ 0.9,
      /* los    */ 4);
```
```c++
   SetUnitBasic(s,
      MaxHp(40),
      MArmor(0), 
      PArmor(1),
      Radius(1),
      Speed(0.9),
      Los(4));
```
```c++
    SetUnitBasic(s,
      MaxHp  (40),
      MArmor (0), 
      PArmor (1),
      Radius (1),
      Speed  (0.9),
      Los    (4));
```

> I know two people is basically impossible to agree in one format

I think we have reached this state

Reply:
> > I know two people is basically impossible to agree in one format
> 
> I think we have reached this state

Then, given that you currently put in the most effort by far, I think that you should decide :)